packages/k6-tests/src/koko/030-search-for-filename.ts

```diff
@@ -0,0 +1,142 @@
+import { check } from 'k6';
+import exec from 'k6/execution';
+import { Options } from 'k6/options';
+import { times } from 'lodash'
+
+interface Credential {
+  login: string;
+  password: string;
+}
+
+interface Info {
+  credential: Credential;
+  home: string;
+}
+
+interface Data {
+  adminCredential: Credential;
+  userInfos: Info[];
+}
+
+interface Settings {
+  authAdapter: Adapter;
+  baseURL: string;
+  clientVersion: Version;
+  adminUser: Credential;
+  assets: {
+    folderCount: number;
+    textDocumentCount: number;
+  };
+  k6: Options;
+}
+
+/**/
+const settings: Settings = {
+  baseURL: __ENV.BASE_URL || 'https://localhost:9200',
+  authAdapter: __ENV.AUTH_ADAPTER == Adapter.basicAuth ? Adapter.basicAuth : Adapter.openIDConnect,
+  clientVersion: Version[ __ENV.CLIENT_VERSION ] || Version.ocis,
+  adminUser: {
+    login: __ENV.ADMIN_LOGIN || 'admin',
+    password: __ENV.ADMIN_PASSWORD || 'admin',
+  },
+  assets: {
+    folderCount: parseInt(__ENV.ASSETS_FOLDER_COUNT) || 2,
+    textDocumentCount: parseInt(__ENV.ASSETS_TEXT_DOCUMENT_COUNT) || 2,
+  },
+  k6: {
+    vus: 1,
+    insecureSkipTLSVerify: true,
+  },
+};
+
+/**/
+export const options: Options = settings.k6;
+
+export function setup(): Data {
+  const adminCredential = settings.adminUser;
+  const adminClient = new Client(settings.baseURL, settings.clientVersion, settings.authAdapter, adminCredential);
+
+  const userInfos = times<Info>(options.vus || 1, () => {
+    const userCredential = { login: randomString(), password: randomString() };
+    adminClient.user.create(userCredential);
+    adminClient.user.enable(userCredential.login);
+
+    const userClient = new Client(settings.baseURL, settings.clientVersion, settings.authAdapter, userCredential);
+    const userDrivesResponse = userClient.user.drives();
+    const [userHome = userCredential.login] = queryJson('$.value[?(@.driveType === \'personal\')].id', userDrivesResponse?.json());
+
+    return {
+      home: userHome,
+      credential: userCredential,
+    };
+  });
+
+  return {
+    adminCredential,
+    userInfos,
+  };
+}
+
+export default function ({ userInfos }: Data): void {
+  const defer: (() => void)[] = [];
+  const { home: userHome, credential: userCredential } = userInfos[ exec.vu.idInTest - 1 ];
+  const userClient = new Client(settings.baseURL, settings.clientVersion, settings.authAdapter, userCredential);
+  const searchTasks = new Queue({ delay: 1000, delayMultiplier: 0 })
+  const folderNames = times(settings.assets.folderCount, () => randomString())
+  const textDocuments = times(settings.assets.textDocumentCount, () => [randomString() + '.txt', randomString()])
+
+  const searchTask = (query: string, expectedId: string, description: string) => {
+    const searchResponse = userClient.search.resource(userCredential.login, { query })
+    const [searchFileID] = queryXml('$..oc:fileid', searchResponse?.body);
+    const ready = !!searchFileID
```

Comment:
i get your point, i'm also not happy with that code part, but i have to explain why it behaves like that and why this is needed.

but first let me explain the `!!`, technically it converts the maybe string value to a boolean, so `!!searchFileID` is the same as `!!""` (false) and `!!"foo"` (true). So it behaves exactly the same in the if condition like a empty or non empty string. Even if it achieves the same result i prefer to check booleans.

ok now let me explain what that code path is doing and why you see the wild output when `--http-debug` is set (same happens without but you do not see it).

for oCis with post-processing and same for oCis without the oCis search service indexing happens at some point in time, no guarantee when it happens. So in theory it could happen that the file is already there but the indexing takes some hours (eventually).

in this test we test the file search, so we need some way to identify when the search index is ready and the resource was indexed.

this is why you see the "wild" outputs without a result, we use Promise recursion with exponential backoff to identify if the resource has been processed.

As i said in the beginning, i'm not happy with that solution, i have ideas to opt into the nats events, use sse to get notifications when the file is ready, or something else (basically everything is better compared to polling)! But then we still have the same problem for `occ` and `nc` if the same processing model applies there.

Sorry for the long explanation, but the background is needed to get a better understanding of the problem.

Can you please try it more often with the `!!searchFileID` in place and check if it always behave the same!? I Know i told `!!searchFileID` is the same as `!!""` or `!!"foo"`, but since k6 uses goja `https://github.com/dop251/goja` and not the native javaScript runtime it could be theoretically.

Reply:
Thanks, that expaint it me. 
I tried to run with `!!searchFileID` many times. about every 10th run has problems
[030Test-log.txt](https://github.com/owncloud/cdperf/files/11157882/030Test-log.txt)


I understand the problem with the search index and am not waiting for a magic fix right now.
just so we know about it.